### ISSUE TITLE
True shared memory for TEE_CheckMemoryAccessRights()

### ISSUE THREAD

**Issue by Hussain1811** [OP-TEE/optee_os#4813]
The TEE internal spec has the below line:

> when true memory sharing is implemented between Clients and the Trusted Application, the Memory
> Reference Parameters passed to the TA entry points will typically not satisfy these requirements. In this case,
> the function TEE_CheckMemoryAccessRights SHALL return TEE_ERROR_ACCESS_DENIED.

Does "true shared memory" mean usage of TEEC_AllocateSharedMemory()/ TEEC_RegisterSharedMemory() functions?
if yes, does the above statement also mean if we pass along this shared mem instance over to the TA, then TEE_CheckMemoryAccessRights() would return TEE_ERROR_ACCESS_DENIED regardless of the access flags?

if not, what does true shared memory actually mean?

**Comment by jenswi-linaro** [comment]
That passage from the specification does look a bit odd. I'd say that with `TEE_MEMORY_ACCESS_ANY_OWNER` you accept that you cannot expect neither Read-after-read consistency, Read-after-write consistency, nor Non-observability. Matching well with the properties of true shared memory. In OP-TEE we assume all non-secure memory is like this regardless of how it's shared.

@jforissier, what's your take on this?

**Comment by jforissier** [comment]
@jenswi-linaro I read it the same as you, especially regarding the "regardless of how it's shared" part, and I hope we got it right in the implementation. Small digression on this aspect... Consider for instance an in/out temporary memory reference, with `CFG_DYN_SHM != y`. In this case, we will copy in/out non-secure memory to/from secure memory when switching worlds. If the TA calls `TEE_CheckMemoryAccessRights()` on the **secure** buffer with `TEE_MEMORY_ACCESS_ANY_OWNER` **not** set, I expect the function to return `TEE_ERROR_ACCESS_DENIED` because although read-after-read and read-after-write consistency would be ensured, non-observability would not (obviously, being an out parameter). I believe it is the case thanks to the use of the `VM_FLAG_EPHEMERAL` flag, isn't it?

**Comment by jenswi-linaro** [comment]
> I believe it is the case thanks to the use of the `VM_FLAG_EPHEMERAL` flag, isn't it?

Correct.

**Comment by jenswi-linaro** [comment]
> Consider for instance an in/out temporary memory reference, with `CFG_DYN_SHM != y`. In this case, we will copy in/out non-secure memory to/from secure memory when switching worlds.

No, if there's any copy it will be when switching between linux user space and kernel space. So the memory may still be tampered with by normal world while the TA is executing.

**Comment by Hussain1811** [comment]
So, if i'm not wrong, as per this discussion, TEE_SUCCESS shouldn't be returned when TEE_MEMORY_ACCESS_ANY_OWNER is set for true shared memory?!?
